src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const updatedState = [];
+  let tempObject = { ...state };
+
+  for (const action of actions) {
+    switch (true) {
+      case action.type === 'addProperties':
+        add(action.extraData);
+        break;
+      case action.type === 'removeProperties':
+        remove(action.keysToRemove);
+        break;
+      case action.type === 'clear':
+        clear();
+        break;
+    }
+  }
+
+  function add(data) {
+    tempObject = { ...tempObject, ...data };
+    updatedState.push({ ...tempObject });
+  }
+
+  function remove(data) {
+    for (const del of data) {
+      delete tempObject[del];
+    }
+    updatedState.push({ ...tempObject });
+  }
+
+  function clear() {
+    tempObject = {};
+    updatedState.push({ ...tempObject });
+  }
+
+  return updatedState;
```

Comment:
Ensure that the original 'state' object is not being modified by any of the operations. As per the task description, the initial 'state' object must not be mutated in any way.